src/useCookieConsent.ts

```diff
@@ -13,21 +13,33 @@ import { allCookiesSetToValue, allPropsApproved } from './utils';
 
 export const useCookieConsent = (
   options?: CookieConsentOptions
-): CookieConsentHookState => {
+): ConsentHookStateType => {
+  const doesStorageExist =
+    typeof options !== 'undefined' && 'storage' in options;
+
+  const storedConsent = doesStorageExist
+    ? JSON.parse(options?.storage.getItem(COOKIE_CONSENT_KEY))
+    : Cookies.getJSON(COOKIE_CONSENT_KEY);
+
   const initialConsent: CookieConsent =
-    Cookies.getJSON(COOKIE_CONSENT_KEY) ||
-    options?.defaultConsent ||
-    DEFAULT_CONSENT;
+    storedConsent || options?.defaultConsent || DEFAULT_CONSENT;
 
   const [consent, setConsent] = useState<CookieConsent>(initialConsent);
 
   useEffect(() => {
     if (consent?.necessary) {
-      Cookies.set(
-        COOKIE_CONSENT_KEY,
-        consent,
-        options?.consentCookieAttributes
-      );
+      if (doesStorageExist) {
+        options.storage.setItem(COOKIE_CONSENT_KEY, JSON.stringify(consent));
+      } else if (
+        typeof options === 'undefined' ||
+        'consentCookieAttributes' in options
```

Comment:
I would say that currently the cookies would be used if the storage is not, so the logical check here would be `!doesStorageExist` also known as jus `else` statement

Reply:
Have to say, your logic makes sense to me. However I have these checks here, for compiler reasons(typescript throws error). I'm unsure how to proceed.

Reply:
You just need to think of when each `if`, `else if`, and `else` scope will execute and think if that covers all the cases that could happen. In this case you have a couple of problems:

1. If neither `if` statements are satisfied, there's a case where nothing happens - and that should never happen.
2. The `if else` check is not correct - the cookies should be set by default, if there are extra options passed - only then they should not be set.

A common approach to making the typescript not complain is to shuffle around the statement after inverting them. For example:

```js
if (foo) {
  // case 1
} else {
  // case 2 {
}
```

Can also be:

```js
if (!foo) {
  // case 2
} else {
  // case 1
}
```

I would also suggest taking a look at "Boolean Algebra" and "Karnaugh Maps". Both sound scarier than they are - they are simple concepts for structuring the way you think about binary (boolean) logic and provide you with some formulas that are just time-consuming to come up with yourself.

Reply:
Fortunately for you, I can't tell you the answer right away, since at that point I could just write it myself, which is not the goal here.